lib/LedgerNodeRecords.js

```diff
@@ -36,9 +37,36 @@ module.exports = class LedgerNodeRecords {
     // ...and if not, call it again and cache its result; if we don't have
     // that block height, we should throw `NotFoundError`
 
+    // use a cache for the latest block summary since this function may
+    // be called often
+    const getLastestBlockSummaryFn = () => this.ledgerNode.storage.blocks
+      .getLatestSummary(this.ledgerNode);
+    let latestBlockSummary = await this._latestBlockSummaryCache.memoize({
+      key: 'latestBlockSummary', fn: getLastestBlockSummaryFn});
+    let latestBlockHeight = latestBlockSummary.eventBlock.block.blockHeight;
+
+    // always set maxBlockHeight since it's used as part of an LRU key
+    if(maxBlockHeight === undefined) {
+      maxBlockHeight = latestBlockHeight;
+    } else if(maxBlockHeight > latestBlockHeight) {
+      // clear the summary cache if `maxBlockHeight` is greater than
+      // `latestBlockHeight` and set `maxBlockHeight` to freshest database value
+      // only if maxBlockHeight is still sless than latestBlockHeight
+      this._latestBlockSummaryCache.delete('latestBlockSummary');
+      latestBlockSummary = await this._latestBlockSummaryCache.memoize({
+        key: 'latestBlockSummary', fn: getLastestBlockSummaryFn});
+      latestBlockHeight = latestBlockSummary.eventBlock.block.blockHeight;
+      if(maxBlockHeight > latestBlockHeight) {
+        maxBlockHeight = latestBlockSummary.eventBlock.block.blockHeight;
+      }
+    }
+
+    // get the record from the LRU cache if possible
     const key = `${maxBlockHeight}|${recordId}`;
-    const fn = () => this._getUncachedRecord({maxBlockHeight, recordId});
-    return this._cache.memoize({key, fn});
+    const getUncachedRecordFn = () => this._getUncachedRecord({
+      maxBlockHeight, recordId});
+
+    return this._cache.memoize({key, fn: getUncachedRecordFn});
```

Comment:
-1 to these changes, AFAICT, all it did was make the code longer and more complicated? It does not look more readable/better in any way to me.

Reply:
The change was made because there were two functions defined, so continuing to use 'fn' wasn't an option. I'll look at cleaning it up by moving the other function definition outside of the code path.

Reply:
Oh, yeah, that makes sense. I went looking for something like that and still missed it somehow before. +1 to refactoring that other bit above so it can be moved back to what it was.